src/osd/objclass.cc

```diff
@@ -700,3 +700,45 @@ uint64_t cls_get_pool_stripe_width(cls_method_context_t hctx)
 
   return ctx->pg->get_pool().stripe_width;
 }
+
+struct GatherFinisher : public PrimaryLogPG::OpFinisher {
+  std::map<std::string, bufferlist> *src_obj_buffs;
+  GatherFinisher(std::map<std::string, bufferlist> *sob) : src_obj_buffs(sob) {}
+  int execute() override {
+    return 0;
+  }
+};
+
+int cls_cxx_gather(cls_method_context_t hctx, const std::set<std::string> &src_objs, const std::string& pool,
+		   const char *cls, const char *method, bufferlist& inbl)
+{
+  PrimaryLogPG::OpContext **pctx = (PrimaryLogPG::OpContext**)hctx;
+  std::map<std::string, bufferlist> *src_obj_buffs = new std::map<std::string, bufferlist>;
```

Comment:
We really try not to do direct allocations like this as it can lead to leaks, like this one does.  GatherFinisher doesn't actually free src_obj_buffs in its destructor, so unless cls_cxx_get_gathered_data runs, we leak the buffers.  Modify GatherFinisher to directly contain src_obj_buffs by value and populate it directly here.  Something like:

```
auto [iter, inserted] = (*pctx->op_finishers).emplace(std::make_pair((*pctx)->current_osd_subop_num, std::make_unique<GatherFinisher>());
assert(inserted);
auto &bufs = iter->second;
for (const auto &obj : src_objs) {
  (bufs)[obj];
}
```

Reply:
The OpContext owns the reference to the GatherFinisher via the unique_ptr and the GatherFinisher owns the buffers directly ensuring that the destruction of the OpContext through whatever means will clean up the buffers.